### ISSUE TITLE
JDK 15: new JDK method means we can't call "isEmpty" on an Array[Char] due to ambiguous implicit conversions and can't compile scala.collection.mutable.StringBuilder

### ISSUE THREAD

**Issue by xuwei-k** [scala/bug#12172]
## reproduction steps

run `sbt compile` in current 2.13.x branch
https://github.com/scala/scala/tree/8c86b7d7136839538cca0ff8fca50f59437564c0

## problem

```
scala/scala/src/library/scala/collection/mutable/StringBuilder.scala:41:13: cannot override a concrete member without a third member that's overridden by both (this rule is designed to prevent ``accidental overrides'')
[error] override def isEmpty: Boolean (defined in trait SeqOps)
[error]   with <defaultmethod> def isEmpty(): Boolean (defined in trait CharSequence)
[error] final class StringBuilder(val underlying: java.lang.StringBuilder) extends AbstractSeq[Char]
[error]             ^
```

`java.lang.CharSequence` has `isEmpty` method since JDK 15

https://github.com/openjdk/jdk/commit/113c48f5da32b26a4f820ccd0e43dae3eba7b73e

```
Welcome to Scala 2.13.3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_201).
Type in expressions for evaluation. Or try :help.

scala> classOf[scala.collection.mutable.StringBuilder].getInterfaces
val res0: Array[Class[_]] = Array(interface scala.collection.mutable.ReusableBuilder, interface scala.collection.mutable.IndexedSeq, interface java.lang.CharSequence, interface java.io.Serializable)
```

**Comment by som-snytt** [comment]
Also 2.12.x:
```
[error] /Users/.../scala/src/library/scala/collection/mutable/StringBuilder.scala:34:13: overriding method isEmpty in trait CharSequence of type ()Boolean;
[error]  method isEmpty in trait IndexedSeqOptimized of type => Boolean cannot override a concrete member without a third member that's overridden by both (this rule is designed to prevent ``accidental overrides'')
[error] final class StringBuilder(private val underlying: JavaStringBuilder)
[error]             ^
[error] one error found
[error] (library / Compile / compileIncremental) Compilation failed
[error] Total time: 44 s, completed Oct 5, 2020, 4:08:52 AM
```

**Comment by catap** [comment]
The issue much deeper than "can't compile" scala-library and that I've "fixed" on my wrongs PRs.

Let assume a code:
```
val array = new Array[Char](0)
array.isEmpty
```

This code can't be compiled on JDK15+ because `isEmpty` can be used from `scala.collection.ArrayOps[Char]` and from `ArrayCharSequence` that inherits `java.lang.CharSequence` with a brand new `isEmpty`.

:(

**Comment by som-snytt** [comment]
Earlier today on dotty:
```
[error] -- [E007] Type Mismatch Error: /Users/.../dotty/compiler/src/dotty/tools/dotc/reporting/MessageRendering.scala:63:32 
[error] 63 |      new String(line) :: { if (rest.isEmpty) Nil else linesFrom(rest) }
[error]    |                                ^^^^
[error]    |Found:    (rest : Array[Char])
[error]    |Required: ?{ isEmpty: ? }
[error]    |Note that implicit extension methods cannot be applied because they are ambiguous;
[error]    |both method ArrayCharSequence in object Predef and method charArrayOps in object Predef provide an extension method `isEmpty` on (rest : Array[Char])
[error] one error found
[error] (dotty-compiler / Compile / compileIncremental) Compilation failed
[error] Total time: 56 s, completed Oct 10, 2020, 5:13:00 PM
```

**Comment by catap** [comment]
@som-snytt yeah and issue **much** deeper:

<img width="969" alt="Screenshot 2020-10-11 at 04 34 46" src="https://user-images.githubusercontent.com/37775/95669066-731c0b00-0b7c-11eb-8cf9-f55a699b7707.png">

I doubt that it is possibly to solve without **renaming** of `ArrayOps.isEmpty` at least at on scala side that brokes tons of code

:(

**Comment by smarter** [comment]
Hmm, so `Predef` defines both `implicit final class ArrayCharSequence(arrayOfChars: Array[Char]) extends CharSequence` and `implicit def charArrayOps(xs: Array[Char]): ArrayOps[Char]          = new ArrayOps(xs)`, I don't see a way to make one conversion have higher specificity than the other while maintaining forward and backward BC, so it seems that a special-case in the compiler will be needed here.

**Comment by smarter** [comment]
@SethTisue I think the mention of JDK 15 in https://docs.scala-lang.org/overviews/jdk-compatibility/overview.html will need a big asterisk until this is addressed.

**Comment by catap** [comment]
@smarter it also brokes `isEmpty` on `scala.reflect.internal.Name`.

**Comment by som-snytt** [comment]
That's why scaladoc uses asterisks. It's asterisks all the way down.
```
/*
 * On JDK15, use `!Array('c').nonEmpty`
 */
```

**Comment by catap** [comment]
@som-snytt at least dotty doesn't follow this remarks :)

**Comment by SethTisue** [comment]
Note that the workaround is to explicitly type-ascribe to `ArrayOps`:

```scala
scala 2.13.3> Array[Char](0).isEmpty
                         ^
              error: type mismatch;
               found   : Array[Char]
               required: ?{def isEmpty: ?}
              Note that implicit conversions are not applicable because they are ambiguous:
               both method ArrayCharSequence in object Predef of type (arrayOfChars: Array[Char]): ArrayCharSequence
               and method charArrayOps in object Predef of type (xs: Array[Char]): scala.collection.ArrayOps[Char]
               are possible conversion functions from Array[Char] to ?{def isEmpty: ?}
                             ^
              error: value isEmpty is not a member of Array[Char]

scala 2.13.3> (Array[Char](0): collection.ArrayOps[Char]).isEmpty
val res4: Boolean = false
```

It seems to me this is serious enough that we ought to do something about it for 2.12.13 and 2.13.4, but I don't think it's necessary to red-flag it on the JDK compatibility page. The problem comes at compile-time, it's not a sneaky runtime failure; and the compat page already links to the `jdk15` label here in this tracker.

**Comment by ebourg** [comment]
Is it possible to also backport the fix to Scala 2.11?

**Comment by SethTisue** [comment]
@ebourg It's highly unlikely there will be any further 2.11 releases. (You could submit a backport PR to the 2.11.x branch and we'd merge it, but it's highly unlikely it would ever actually be officially released. Users are free to publish their own Scala locally.)
